### ISSUE TITLE
bug(forge): accessing array elements breaks coverage for if statements

### ISSUE THREAD

**Issue by PaulRBerg** [foundry-rs/foundry#4310]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (249538f 2023-02-08T00:12:05.805004Z)

### What command(s) is the bug in?

forge coverage

### Operating System

macOS (Apple Silicon)

### Describe the bug

The mere existence of an array accessor within an if statement breaks the coverage for the said if.

I think that the priority for this issue should be set very high, because it may affect lot of Foundry users.

### Reproduction Steps

Take the following unsuspecting contract:

```solidity
contract Foo {
    function countOdd(uint256[] memory arr) external pure returns (uint256 count) {
        uint256 length = arr.length;
        for (uint256 i = 0; i < length; ++i) {
            if (arr[i] % 2 == 1) {
                count++;
                arr[0];
            }
        }
    }
}
```

And the following test file (hidden by default for brevity reasons, click the toggle below to collapse it):

<details>
<summary>Click me to toggle the tests</summary>

```solidity
import { Test } from "forge-std/Test.sol";

contract FooTest is Test {
    Foo internal foo = new Foo();

    function test_CountOdd_1() external {
        uint256[] memory arr = new uint256[](3);
        arr[0] = 78;
        arr[1] = 493;
        arr[2] = 700;
        uint256 count = foo.countOdd(arr);
        assertEq(count, 1);
    }

    function test_CountOdd_2() external {
        uint256[] memory arr = new uint256[](4);
        arr[0] = 78;
        arr[1] = 493;
        arr[2] = 700;
        arr[3] = 1729;
        uint256 count = foo.countOdd(arr);
        assertEq(count, 2);
    }
}

```

</details>

Now, run `forge coverage`. You will get this report:

```text
| File        | % Lines       | % Statements  | % Branches   | % Funcs       |
|-------------|---------------|---------------|--------------|---------------|
| src/Foo.sol | 100.00% (4/4) | 100.00% (6/6) | 50.00% (1/2) | 100.00% (1/1) |
| Total       | 100.00% (4/4) | 100.00% (6/6) | 50.00% (1/2) | 100.00% (1/1) |
```

Notice that the branch coverage is 50% instead of 100%, even if we have tests that account for the full gamut of possibilities (`arr[i]` odd and even).

Now, do this - comment the `arr[0]` expression, and the branch coverage will magically increase to 100%:

```text
| File        | % Lines       | % Statements  | % Branches    | % Funcs       |
|-------------|---------------|---------------|---------------|---------------|
| src/Foo.sol | 100.00% (4/4) | 100.00% (6/6) | 100.00% (2/2) | 100.00% (1/1) |
| Total       | 100.00% (4/4) | 100.00% (6/6) | 100.00% (2/2) | 100.00% (1/1) |

```

Finally, here's an even juicier example:

```solidity
contract Foo {
    function isOdd(uint256 number) external view returns (bool odd) {
        uint256[] memory arr = new uint256[](1);
        arr[0] = block.timestamp;

        if (number % 2 == 1) {
            odd = true;
            arr[0];
        } else {
            odd = false;
        }
    }
}
```

Yet again, the mere existence of the `arr[0]` expression breaks the coverage for the `if (number % 2 == 1)` line. This example is even more surprising because the `arr` array has nothing to do with the actual logic of the function.

### Related Issues

- This bug may be the same as #4309 (this other issue may be a special case of the bug explained in this issue).
- This bug may have also played a role in #3600 (specifically, the second incorrectly reported line in KittyFu307's code).

**Comment by frontier159** [comment]
Fantastic sleuthing @PaulRBerg, much appreciated

**Comment by PaulRBerg** [comment]
Thanks, @frontier159.

After a few hours of additional debugging, I now realize that this issue might be a special case of the other issue that I have just opened, i.e. https://github.com/foundry-rs/foundry/issues/4311.

Though I am not sure - `arr[0]` loads from memory not from the stack 🤔

**Comment by PaulRBerg** [comment]
Note: this bug happens even in cases when it is impossible to get an "index out of bounds" error. I have already shared an example in the OP above, but here's another from my real project:

<img width="1049" alt="Screenshot 2023-02-09 at 6 54 37 PM" src="https://user-images.githubusercontent.com/8782666/217883428-2b5d21fb-825d-4e4b-ae0a-0ae6128614b1.png">

The if statement checks if the `index` is at least 2, so the expression `index - 2` cannot possibly cause an error.

**Comment by mountainpath9** [comment]
Based upon some analysis using the report implemented in #6563, I think this is due to assumptions in the heuristics used to establish which evm instructions correspond to which branch points. This heuristic is described here:

https://github.com/foundry-rs/foundry/blob/0ae39ea8e43e629d67e9373d189f3730019b4bfc/crates/evm/coverage/src/anchors.rs#L72

and, I suspect will only work in some cases.
